server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java

```diff
@@ -291,12 +274,11 @@ public void addTimeoutListener(@Nullable final TimeValue timeout, final TimeoutC
             threadPoolExecutor.execute(new SourcePrioritizedRunnable(Priority.HIGH, "_add_listener_") {
                 @Override
                 public void run() {
+                    final NotifyTimeout notifyTimeout = new NotifyTimeout(listener, timeout);
                     if (timeout != null) {
-                        NotifyTimeout notifyTimeout = new NotifyTimeout(listener, timeout);
                         notifyTimeout.cancellable = threadPool.schedule(notifyTimeout, timeout, ThreadPool.Names.GENERIC);
```

Comment:
Is there not a race here in that the scheduled task could run before adding the listener to the map below and the scheduled task could have called `removeTimeoutListener` before it is put into the map?

I recognize this preserves existing behavior, but if you agree I think it is worth fixing in this PR.

Reply:
Yea it's a trivial fix I think, I'll clean it up here.